### PR TITLE
Port C API docs to Meson

### PR DESCRIPTION
**Pull request by fricklerhandwerk**
Thanks @Ericson2314 for the introduction to the process

# Motivation
https://github.com/NixOS/nix/issues/2503

# Context
previously: #10930

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: package.nix [hunk 1]
```diff
@@ -213,6 +201,7 @@ in {
     autoconf-archive
     autoreconfHook
     pkg-config
+    doxygen
```
Comment (Ericson2314) [inline package.nix:204]: Is this for the dev shell? I think we can do in another way fwiw in the `flake.nix`, where I combine the deps of a few derivations already

File: src/external-api-docs/package.nix [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  version = lib.fileContents ./.version + versionSuffix;
+
+  src = fileset.toSource {
+    root = ../..;
+    fileset = let
+      cpp = fileset.fileFilter (file: file.hasExt "cc" || file.hasExt "h");
+    in fileset.unions [
+      ./meson.build
+      ./doxygen.cfg.in
+      ./README.md
```
Comment (Ericson2314) [inline src/external-api-docs/package.nix:26]: Do we need this?

Comment (Ericson2314) [reply src/external-api-docs/package.nix:26]: Oh yes we do

**Comment by fricklerhandwerk** [comment]
@Ericson2314 For some reason this separate derivation builds all of Nix, but the internal API docs don't.

File: package.nix (hunk 1 above)
Comment (fricklerhandwerk) [reply package.nix]: Yes, but indeed we may want to rework this later.

**Comment by Ericson2314** [comment]
Are you sure? I did not experience that I don't think.

File: src/external-api-docs/meson.build [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+fs = import('fs')
+
+doxygen_cfg = configure_file(
+  input : 'doxygen.cfg.in',
+  output : 'doxygen.cfg',
+  configuration : {
+    'PROJECT_NUMBER': meson.project_version(),
+    'OUTPUT_DIRECTORY' : meson.current_build_dir(),
+    #'INCLUDE_PATH' : meson.current_build_dir(),
```
Comment (fricklerhandwerk) [inline src/external-api-docs/meson.build:15]:
```suggestion
```

**Comment by edolstra** [comment]
Can we get rid of flake outputs like `packages.x86_64-linux.nix-external-api-docs-armv6l-unknown-linux-gnueabihf `? We only need one `packages.x86_64-linux.nix-external-api-docs`.

**Pull request merged**